tests/test_provenance.py

```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
         self.assertEqual(decoded.parameters.start_time, 0)
         self.assertEqual(decoded.parameters.end_time, 100)
         self.assertEqual(decoded.parameters.keep, False)
+
+
+class TestParseProvenance(unittest.TestCase):
+    def test_parse_provenance(self):
+        # Most testing of parsing is in the round tripping below
+        with self.assertRaises(ValueError):
+            ts = msprime.simulate(5)
+            prov = json.loads(ts.provenance(0).record)
+            prov['software']['name'] = 'skynet'
+            tables = ts.tables
+            tables.provenances.clear()
+            tables.provenances.add_row(json.dumps(prov))
+            command, prov = msprime.provenance.parse_provenance(tables.tree_sequence())
+
+
+class TestRoundTrip(unittest.TestCase):
+    """
+    Check that we can recreate tree sequences from their provenances
+    """
+
+    def verify_equal(self, ts1, ts2):
+        ts1_tables = ts1.dump_tables()
+        ts2_tables = ts2.dump_tables()
+        # We clear the provenances as they are not equal due to timestamps
+        ts1_tables.provenances.clear()
+        ts2_tables.provenances.clear()
+        self.assertEqual(ts1_tables, ts2_tables)
+
+    def test_simulate_round_trip(self):
+        pop_configs = [msprime.PopulationConfiguration(5) for _ in range(2)]
+        ts = msprime.simulate(population_configurations=pop_configs,
+                              migration_matrix=[[0, 1], [1, 0]],
+                              demographic_events=[msprime.SimulationModelChange(
+                                  time=lambda t: t+0.1),
+                              ])
+        command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
+        recreated = getattr(msprime, command)(**parsed_prov)
+        self.verify_equal(ts, recreated)
+
```

Comment:
Nice to split these up into discrete tests. First one is ``test_population_structure``, second is ``test_pedigree``, etc. Also nice to check a few more simple cases, like
```
def verify(ts):
    command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
    recreated = getattr(msprime, command)(**parsed_prov)
    self.verify_equal(ts, recreated)


def test_sample_size(self):
    ts = msprime.simulate(10)
    self.verify(ts)

def test_random_seed(self):
     ts = msprime.simulate(10, random_seed=1234)
     self.verify(ts)
```

No need to be exhaustive, but good to capture a few of common usages here.

Also, good to test if we can round-round-trip --- we're not checking the provenance of the tree sequences we're generating from the parsed provenance. Can do this as a separate test I think.

Reply:
Nice catch - the round-round-trip found an edge case with functional arguments.